### PR TITLE
rbd-nbd: support notrim option with map command

### PR DESCRIPTION
**Pull request by pkalever**
currently --notrim option works for rbd kernel mounter, but fails with rbd-nbd

```
$ rbd device --options notrim map rbd-pool/image0
/dev/rbd0
$ rbd device list
id  pool      namespace  image   snap  device
0   rbd-pool             image0  -     /dev/rbd0

$ rbd device --device-type nbd --options try-netlink,notrim map rbd-pool/image0
rbd-nbd: unknown args: --notrim
rbd: rbd-nbd failed with error: /data/ceph/build/bin/rbd-nbd: exit status: 1
```
With these changes:
```
$ rbd device --device-type nbd --options try-netlink,notrim map rbd-pool/image0
/dev/nbd0
$ rbd-nbd list-mapped
id    pool      namespace  image   snap  device
6945  rbd-pool             image0  -     /dev/nbd0
$ ps -eo "cmd" |grep [r]bd-nbd
/data/ceph/build/bin/rbd-nbd map rbd-pool/image0 --try-netlink --notrim
```

Fixes: https://github.com/ceph/ceph-csi/issues/2171

Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

### CODE REVIEW

**Comment by pkalever** [comment]
cc: @nixpanic

**Review by nixpanic — COMMENTED** [review]
Thanks, looks like a reasonable changes to me :+1:

**Review by trociny — APPROVED** [review]
LGTM

**Review by idryomov — CHANGES_REQUESTED** [review]
Looks good to me but need to add a test case to `qa/workunits/rbd/rbd-nbd.sh`.  There already is a "trim test" one, add a counter example based on it.

File: qa/workunits/rbd/rbd-nbd.sh [hunk 1]
```diff
@@ -166,15 +166,37 @@ get_pid ${POOL}
...
+# trim test
+DEV=`_sudo rbd device --device-type nbd map ${POOL}/${IMAGE}`
+get_pid ${POOL}
+provisioned=`rbd -p ${POOL} --format xml du ${IMAGE} |
+  $XMLSTARLET sel -t -m "//stats/images/image/provisioned_size" -v .`
+used=`rbd -p ${POOL} --format xml du ${IMAGE} |
+  $XMLSTARLET sel -t -m "//stats/images/image/used_size" -v .`
+[ "${used}" -eq "${provisioned}" ]
+# should honor discard as at time of mapping trim was considered by default
+_sudo mkfs.ext4 -F -E discard ${DEV} # better idea?
```
Comment (idryomov) [inline qa/workunits/rbd/rbd-nbd.sh:198]:
```suggestion
_sudo blkdiscard ${DEV}
```

File: qa/workunits/rbd/rbd-nbd.sh [hunk 2]
```diff
@@ -166,15 +166,37 @@ get_pid ${POOL}
...
+# notrim test
+DEV=`_sudo rbd device --device-type nbd --options notrim map ${POOL}/${IMAGE}`
+get_pid ${POOL}
 provisioned=`rbd -p ${POOL} --format xml du ${IMAGE} |
   $XMLSTARLET sel -t -m "//stats/images/image/provisioned_size" -v .`
 used=`rbd -p ${POOL} --format xml du ${IMAGE} |
   $XMLSTARLET sel -t -m "//stats/images/image/used_size" -v .`
 [ "${used}" -eq "${provisioned}" ]
+# should ignore discard option as at time of mapping notrim was used
 _sudo mkfs.ext4 -E discard ${DEV} # better idea?
```
Comment (idryomov) [inline qa/workunits/rbd/rbd-nbd.sh:180]:
```suggestion
# should fail discard as at time of mapping notrim was used
expect_false _sudo blkdiscard ${DEV}
```

**Review by idryomov — CHANGES_REQUESTED** [review]
A better idea ;)

**Comment by pkalever** [comment]
> A better idea ;)

Indeed :-)

Thanks, @idryomov, refreshed with the suggested changes now.

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
https://pulpito.ceph.com/dis-2021-06-30_11:33:22-rbd-wip-dis-testing-distro-basic-smithi/

**Pull request merged**